File: zenscript-language-server/src/main/java/raylras/zen/lsp/provider/data/Snippet.java
package raylras.zen.lsp.provider.data;

import org.eclipse.lsp4j.*;
import org.eclipse.lsp4j.jsonrpc.messages.Either;
import raylras.zen.model.CompilationEnvironment;
import raylras.zen.model.parser.ZenScriptParser.MemberAccessExprContext;
import raylras.zen.model.symbol.Operator;
import raylras.zen.model.type.*;
import raylras.zen.util.Operators;
import raylras.zen.util.Ranges;

@FunctionalInterface
public interface Snippet {

    CompletionItem get();

    static Snippet dotFor(Type type, CompilationEnvironment env, MemberAccessExprContext ctx) {
        Type iteratorType = Operators.getUnaryResult(type, Operator.FOR_IN, env).orElse(null);
        if (iteratorType instanceof MapType) {
            return () -> createMemberAccessCompletion("for", "for key, value in map", "for ${1:key}, ${2:value} in %s {\n\t$0\n}", ctx);
        } else if (iteratorType instanceof ListType) {
            return () -> createMemberAccessCompletion("for", "for element in list", "for ${1:value} in %s {\n\t$0\n}", ctx);
        } else {
            return () -> null;
        }
    }

    static Snippet dotForI(Type type, CompilationEnvironment env, MemberAccessExprContext ctx) {
        Type iteratorType = Operators.getUnaryResult(type, Operator.FOR_IN, env).orElse(null);
        if (iteratorType instanceof ListType) {
            return () -> createMemberAccessCompletion("fori", "for index, element in list", "for ${1:i}, ${2:value} in %s {\n\t$0\n}", ctx);
        } else {
            return () -> null;
        }
    }

    static Snippet dotVal(MemberAccessExprContext ctx) {
        return () -> createMemberAccessCompletion("val", "val name = expr", "val ${1:value} = %s;", ctx);
    }

    static Snippet dotVar(MemberAccessExprContext ctx) {
        return () -> createMemberAccessCompletion("var", "var name = expr", "var ${1:value} = %s;", ctx);
    }

    static Snippet dotIfNull(Type type, MemberAccessExprContext ctx) {
        if (type instanceof NumberType || type == BoolType.INSTANCE || type == VoidType.INSTANCE) {
            return () -> null;
        } else {
            return () -> createMemberAccessCompletion("null", "if (isNull(expr))", "if (isNull(%s)) {\n\t$0\n}", ctx);
        }
    }

    static Snippet dotIfNotNull(Type type, MemberAccessExprContext ctx) {
        if (type instanceof NumberType || type == BoolType.INSTANCE || type == VoidType.INSTANCE) {
            return () -> null;
        } else {
            return () -> createMemberAccessCompletion("nn", "if (!isNull(expr))", "if (!isNull(%s)) {\n\t$0\n}", ctx);
        }
    }

    private static CompletionItem createMemberAccessCompletion(String name, String description, String snippet, MemberAccessExprContext ctx) {
        CompletionItem item = new CompletionItem(name);
        item.setInsertTextMode(InsertTextMode.AdjustIndentation);
        item.setInsertTextFormat(InsertTextFormat.Snippet);
        CompletionItemLabelDetails labelDetails = new CompletionItemLabelDetails();
        labelDetails.setDescription(description);
        item.setLabelDetails(labelDetails);
        TextEdit textEdit = new TextEdit(Ranges.toLspRange(ctx), snippet.formatted(ctx.expression().getText()));
        item.setTextEdit(Either.forLeft(textEdit));
        item.setSortText(name);
        item.setFilterText(ctx.expression().getText() + "." + name);
        return item;
    }

}

File: zenscript-language-server/src/main/kotlin/raylras/zen/lsp/provider/data/Snippet.kt
package raylras.zen.lsp.provider.data

import org.eclipse.lsp4j.*
import org.eclipse.lsp4j.jsonrpc.messages.Either
import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.parser.ZenScriptParser.MemberAccessExprContext
import raylras.zen.model.symbol.Operator
import raylras.zen.model.symbol.applyUnaryOperator
import raylras.zen.model.type.ListType
import raylras.zen.model.type.MapType
import raylras.zen.model.type.Type
import raylras.zen.util.textRange
import raylras.zen.util.toLspRange

object Snippet {
    fun dotFor(type: Type, env: CompilationEnvironment, ctx: MemberAccessExprContext): CompletionItem? {
        return when (type.applyUnaryOperator(Operator.FOR_IN, env)) {
            is MapType -> {
                createMemberAccessCompletion(
                    "for",
                    "for key, value in map",
                    "for \${1:key}, \${2:value} in %s {\n\t$0\n}",
                    ctx
                )
            }

            is ListType -> {
                createMemberAccessCompletion(
                    "for",
                    "for element in list",
                    "for \${1:value} in %s {\n\t$0\n}",
                    ctx
                )
            }

            else -> null
        }
    }

    fun dotForI(type: Type, env: CompilationEnvironment, ctx: MemberAccessExprContext): CompletionItem? {
        return when (type.applyUnaryOperator(Operator.FOR_IN, env)) {
            is ListType -> {
                createMemberAccessCompletion(
                    "fori",
                    "for index, element in list",
                    "for \${1:i}, \${2:value} in %s {\n\t$0\n}",
                    ctx
                )
            }

            else -> null
        }
    }

    fun dotVal(ctx: MemberAccessExprContext): CompletionItem {
        return createMemberAccessCompletion("val", "val name = expr", "val \${1:value} = %s;", ctx)
    }

    fun dotVar(ctx: MemberAccessExprContext): CompletionItem {
        return createMemberAccessCompletion("var", "var name = expr", "var \${1:value} = %s;", ctx)
    }

    fun dotIfNull(type: Type, ctx: MemberAccessExprContext): CompletionItem? {
        return if (type.isNullable()) {
            createMemberAccessCompletion(
                "null",
                "if (isNull(expr))",
                "if (isNull(%s)) {\n\t$0\n}",
                ctx
            )
        } else {
            null
        }
    }

    fun dotIfNotNull(type: Type, ctx: MemberAccessExprContext): CompletionItem? {
        return if (type.isNullable()) {
            createMemberAccessCompletion(
                "nn",
                "if (!isNull(expr))",
                "if (!isNull(%s)) {\n\t$0\n}",
                ctx
            )
        } else {
            null
        }
    }

    private fun createMemberAccessCompletion(
        name: String,
        description: String,
        snippet: String,
        ctx: MemberAccessExprContext
    ): CompletionItem {
        val item = CompletionItem(name)
        item.insertTextMode = InsertTextMode.AdjustIndentation
        item.insertTextFormat = InsertTextFormat.Snippet
        val labelDetails = CompletionItemLabelDetails()
        labelDetails.description = description
        item.labelDetails = labelDetails
        val textEdit = TextEdit(ctx.textRange.toLspRange(), snippet.format(ctx.expression().text))
        item.textEdit = Either.forLeft(textEdit)
        item.sortText = name
        item.filterText = ctx.expression().text + "." + name
        return item
    }
}
